docs/design/services/dynamodb/high-level-library/prototype/option-1/sync/Prototype.java

```diff
@@ -0,0 +1,1026 @@
+     */
+    ItemAttributeValueType type();
+
+    /**
+     * The {@code is*} methods can be used to check the underlying DynamoDB-specific type of the attribute value.
+     *
+     * If the type isn't known (eg. because it was created via {@link ItemAttributeValue#from(Object)}), {@link #isJavaType()}
+     * will return true. Such types will be converted into DynamoDB-specific types by the document client before they are
+     * persisted.
+     */
+
+    boolean isItem();
+    boolean isString();
+    boolean isNumber();
+    boolean isBytes();
+    boolean isBoolean();
+    boolean isListOfStrings();
+    boolean isListOfNumbers();
+    boolean isListOfBytes();
+    boolean isListOfAttributeValues();
+    boolean isNull();
+    boolean isJavaType();
+
+    /**
+     * The {@code as*} methods can be used to retrieve this value without the overhead of type conversion of {@link #as(Class)}.
+     *
+     * An exception will be thrown from these methods if the requested type does not match the actual underlying type. When
+     * the type isn't know, the {@code is*} or {@link #type()} methods can be used to query the underlying type before
+     * invoking these {@code as*} methods.
+     */
+    Item asItem();
+    String asString();
+    BigDecimal asNumber();
+    SdkBytes asBytes();
+    Boolean asBoolean();
+    List<String> asListOfStrings();
+    List<BigDecimal> asListOfNumbers();
+    List<SdkBytes> asListOfBytes();
+    List<ItemAttributeValue> asListOfAttributeValues();
+    Object asJavaType();
+
+    /**
+     * Convert this attribute value from a {@link ItemAttributeValueType#JAVA_TYPE} to a type that can be persisted in DynamoDB.
+     *
+     * This will throw an exception if {@link #isJavaType()} is false.
+     */
+    ItemAttributeValue convertFromJavaType(ItemAttributeValueConverter<?> converter);
+}
+
+/**
+ * The schema for a specific item. This describes the item's structure and which attributes it contains.
+ *
+ * This is mostly an implementation detail, and can be ignored except by developers interested in creating
+ * {@link ItemAttributeValueConverter}.
+ */
+@ThreadSafe
+public interface ItemSchema {
+    /**
+     * Create a builder for configuring and creating an {@link ItemSchema}.
+     */
+    static ItemSchema.Builder builder();
+
+    interface Builder {
+        /**
+         * Specify the attribute schemas that describe each attribute of this item.
+         */
+        ItemSchema.Builder attributeSchemas(Map<String, ItemAttributeSchema> attributeSchemas);
+        ItemSchema.Builder putAttributeSchema(String attributeName, ItemAttributeSchema attributeSchema);
+        ItemSchema.Builder removeAttributeSchema(String attributeName);
+        ItemSchema.Builder clearAttributeSchemas();
+
+        /**
+         * The converter that should be used for converting all items that conform to this schema.
+         */
+        ItemSchema.Builder converter(ItemAttributeValueConverter<?> converter);
```

Comment:
Was this meant to be a different type? Seems like `ItemAttributeValueConverter` is for individual AttributeValues

Reply:
The syntax may not be perfect. I hand-waved it since I figured we could figure out the exact details at implementation time.

This is the converter to use for converting this specific Item into an ItemAttributeValue, assuming it's not overridden at the attribute level. 

Consider the following class structure:
```Java
class Car {
    @Attribute
    private Engine engine;
}

@Converter(EngineConverter.class)
class Engine {
    @Attribute
    private int cylinderCount;
}
```

We'd have the following ItemSchemas:
```
{
    javaType: Car.class,
    attributeSchemas: {
        engine: {
            javaType: Engine.class
        }
    }
}

{
    javaType: Engine.class,
    converter: EngineConverter.class,
    attributeSchemas: {
        cylinderCount: {
            javaType: int.class
        }
    }
}
```

This would instruct the SDK to marshal all usages of this specific `Engine` type using `EngineConverter`.